File: script/make/protocol/js/ShopProtocol.js
export class ShopQueryRequest {
    /** @type {number} protocol **/
    protocol = 11301;
    /**
     * @type {{
     * }} data
    **/
    data;
}

export class ShopQueryResponse {
    /** @type {number} protocol **/
    protocol = 11301;
    /**
     * @type {Array<{
     *     shopId: number;                                                                          // 商店ID
     *     number: number;                                                                          // 数量
     * }>} data
    **/
    data;
}

export class ShopBuyRequest {
    /** @type {number} protocol **/
    protocol = 11302;
    /**
     * @type {{
     *     shopId: number;                                                                          // 商店ID
     *     number: number;                                                                          // 数量
     * }} data
    **/
    data;
}

export class ShopBuyResponse {
    /** @type {number} protocol **/
    protocol = 11302;
    /**
     * @type {string} data
    **/
    data;
}

export default class ShopProtocol {
    static encode(textEncoder, view, offset, protocol, data) {
        switch (protocol) {
            case 11301: {

                return new DataView(view.buffer.slice(0, offset));
            }
            case 11302: {

                // extend
                while (view.byteLength < offset + 4) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 商店ID
                view.setUint32(offset, data.shopId, false);
                offset = offset + 4;
                // extend
                while (view.byteLength < offset + 2) {
                    const extendView = new DataView(new ArrayBuffer(view.byteLength * 2));
                    (new Uint8Array(extendView.buffer)).set(new Uint8Array(view.buffer));
                    view = extendView;
                }
                // 数量
                view.setUint16(offset, data.number, false);
                offset = offset + 2;
                return new DataView(view.buffer.slice(0, offset));
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }

    static decode(textDecoder, view, offset, protocol) {
        switch (protocol) {
            case 11301: {
                // 已购买列表
                const data = [];
                let dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                while (--dataLength >= 0) {
                    // 
                    // 商店ID
                    const dataDataShopId = view.getUint32(offset, false);
                    offset = offset + 4;
                    // 数量
                    const dataDataNumber = view.getUint16(offset, false);
                    offset = offset + 2;
                    // object
                    const dataData = {"shopId": dataDataShopId, "number": dataDataNumber};
                    // add
                    data.push(dataData);
                }
                return {"protocol": 11301, "data": data};
            }
            case 11302: {
                // 结果
                const dataLength = view.getUint16(offset, false);
                offset = offset + 2;
                const dataArray = new Uint8Array(view.buffer.slice(offset, offset + dataLength));
                const data = textDecoder.decode(dataArray);
                offset = offset + dataLength;
                return {"protocol": 11302, "data": data};
            }
            default: throw("unknown protocol define: " + protocol)
        }
    }
}

File: script/make/protocol/cs/ShopProtocol.cs
public class ShopQueryRequest
{
    public System.UInt16 protocol = 11301;
    public Empty data;
}

public class ShopQueryResponse
{
    public System.UInt16 protocol = 11301;
    public System.Collections.Generic.List<(
        System.UInt32 shopId,                                                   // 商店ID
        System.UInt16 number                                                    // 数量
    )> data;
}

public class ShopBuyRequest
{
    public System.UInt16 protocol = 11302;
    public (
        System.UInt32 shopId,                                                   // 商店ID
        System.UInt16 number                                                    // 数量
    ) data;
}

public class ShopBuyResponse
{
    public System.UInt16 protocol = 11302;
    public System.String data;
}

public static class ShopProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Object dataRaw) 
    {
        switch (protocol) 
        {
            case 11301:
            {
                #pragma warning disable
                var data = (Empty)dataRaw;
                #pragma warning restore

                return;
            }
            case 11302:
            {
                var data = ((System.UInt32 shopId, System.UInt16 number))dataRaw;

                // 商店ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int32)(System.UInt32)data.shopId));
                // 数量
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int16)(System.UInt16)data.number));
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Object Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 11301:
            {
                // 已购买列表
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = new System.Collections.Generic.List<(System.UInt32 shopId, System.UInt16 number)>(dataLength);
                while (dataLength-- > 0)
                {
                    // 
                    // 商店ID
                    var dataDataShopId = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 数量
                    var dataDataNumber = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    // object
                    var dataData = (shopId: dataDataShopId, number: dataDataNumber);
                    // add
                    data.Add(dataData);
                }
                return (protocol: 11301, data: data);
            }
            case 11302:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return (protocol: 11302, data: data);
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}